### ISSUE TITLE
Playing WaveOut through call?

### ISSUE THREAD

**Issue by alinoob** [sipsorcery/sipsorcery#241]
I'm trying to create a waveout containing text to speech and play the waveout while call is established and then call is ended when waveout is finished. Do you know if that is in someway possible and could you reference to some useful sources in the doc?

thanks in advance

**Comment by sipsorcery** [comment]
Do you mean you want to play an audio file as early media to a caller and then reject the call? Or do you want to answer the call, play the audio file and then hangup?

**Comment by alinoob** [comment]
I want to call someone, play the audio file and then hangup the call.

**Comment by sipsorcery** [comment]
The [PlaySounds](https://github.com/sipsorcery/sipsorcery/blob/master/examples/SIPExamples/PlaySounds/Program.cs) example does that. The audio format needs to be PCM sampled at 8K or 16k.

This [Azure Example](https://github.com/sipsorcery/sipsorcery/tree/master/examples/AzureExamples/TextToPcm) shows how to get a text-to-speech sample in the correct format.

**Comment by alinoob** [comment]
Are these examples also working on .NET Framework?

**Comment by sipsorcery** [comment]
The examples mentioned target .Net Core but they should work fine if you change the target to .Net Framework 4.6 or greater.

**Comment by alinoob** [comment]
Is there a possibility to send better quality audios?

**Comment by sipsorcery** [comment]
You can use the G722 wideband codec. It has a higher frequency range than G711 and does sound a lot less tinny.

**Comment by alinoob** [comment]
Do you have sample code by any chance? NAudio documentation on that part is really poor..

**Comment by sipsorcery** [comment]
You can test G722 with the PlaySounds example without having to change any code. You have two choices to force G722 to be used for the call:

 - On the SIP device you are calling remove all codecs except G722.
 - Or edit the code in the [WindowsAudioRtpSession](https://github.com/sipsorcery/sipsorcery/blob/1f99b45ebe8c5a540d3dbaa6ed43687d828f026a/examples/SIPExamples/Common/Devices/WindowsAudioRtpSession.cs#L57) and remove all the codec options except G722 (if the SIP device you are calling doesn't support G722 the call will fail).

**Comment by sipsorcery** [comment]
Feel free to re-open if necessary.

**Comment by alinoob** [comment]
So, I've been trying this code and the output quality is just really bad. I used this with the PlaySounds example on the latest versions of SIPSorcery:
```
using (SpeechSynthesizer ss = new SpeechSynthesizer())
using (MemoryStream stream = new MemoryStream())
            {
                ss.SetOutputToWaveStream(stream);
                ss.Speak("Testmessage");
                stream.Seek(0, SeekOrigin.Begin);
                var reader = new WaveFileReader(stream);
                var desiredOutputFormat = new WaveFormat(16000, 16, 1);
                var converter = new WaveFormatConversionStream(desiredOutputFormat, reader);

                WaveFileWriter.CreateWaveFile(Path.Combine(AppContext.BaseDirectory, "message.raw"), converter);
            }
````
Is there any way of improving the quality? G722 does not work, since most of telephones (atleast in Germany) does not support it - not sure why either..

**Comment by sipsorcery** [comment]
.NET Framework or Core?

**Comment by alinoob** [comment]
.NET Framework

**Comment by sipsorcery** [comment]
The timing resolution in the .NET Framework for callbacks is not as good as .NET Core.

````csharp
using System;
using System.Threading;

namespace test
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Timer callback test");

            DateTime start = DateTime.Now;
            TimerCallback callback = new TimerCallback((state) =>
            {
                double period = DateTime.Now.Subtract(start).TotalMilliseconds;
                Console.WriteLine("{0}ms", period);
                start = DateTime.Now;
            });

            Timer timer = new Timer(callback, null, 0, 20);

            Console.WriteLine("Press any key to exit.");
            Console.ReadLine();
        }
    }
}
````

**.NET Framework**
````
Timer callback test
Press any key to exit.
11.0018ms
27.0001ms
31.9983ms
31.0032ms
32.0269ms
32.0419ms
30.9421ms
31ms
31.9989ms
31.0007ms
31.9989ms
30.0098ms
30.4215ms
32.0005ms
31.9987ms
30.9944ms
````

**.NET Core**
````
Timer callback test
Press any key to exit.
12.6272ms
17.6951ms
14.8114ms
30.2477ms
19.2122ms
25.617ms
15.9496ms
14.9122ms
14.1295ms
15.2643ms
30.2757ms
14.5772ms
15.4979ms
30.4258ms
14.5932ms
15.0602ms
31.8337ms
14.5095ms
15.9235ms
14.9602ms
````

Ideally VoIP audio sources should use a period of `20ms` which is what the [AudioExtraSource](https://github.com/sipsorcery/sipsorcery/blob/2fec2a0c5ed052c418712e57a115f258b6a0a2cb/src/app/Media/Sources/AudioExtrasSource.cs#L92) uses. You can try tweaking the `20ms` to a larger value to see if you can get adequate performance with the .NET Framework. I never really had much luck with smooth audio from the .NET Framework unless the timing source was external such as a microphone or network.

**Comment by sipsorcery** [comment]
The consequence of larger sample times is that the loss of any audio UDP packet is more noticeable. When you lose packets it will usually sound like a series of clicks or breaks in the audio stream.

I don't know what the noise on initialisation could be. I haven't experienced that.

**Comment by alinoob** [comment]
Update: After trying different values for an hour, I found out that
`private const int AUDIO_SAMPLE_PERIOD_MILLISECONDS = 31;`
is the best value for .NET Framework, atleast on my machine.
Are you planning to change it somehow for future releases on .NET Framework or do I need to compile it myself? :)

After updating, the initialization noise was not gone but way lower, almost not noticable.

**Comment by sipsorcery** [comment]
> Are you planning to change it somehow for future releases on .NET Framework or do I need to compile it myself? :)

I'll make it a configurable value in the next nuget package release.

**Comment by sipsorcery** [comment]
[AudioSamplePeriodMilliseconds](https://github.com/sipsorcery/sipsorcery/blob/c173ac01a1d0399d5bf252a61339b457eb62246e/src/app/Media/Sources/AudioExtrasSource.cs#L166) now available in nuget package version `4.0.82-pre`.
